### ISSUE TITLE
Silent refresh when user login changes

### ISSUE THREAD

**Issue by florisrobbemont** [damienbod/angular-auth-oidc-client#562]
Hi all,

I have a question to which I couldn't find an answer in the docs or the other issues.
We have built a system wich supports impersonation. Backoffice users can start an impersonation session (will always open a new tab) which will then login the user (IdentityServer4). After the login the backoffice user will be redirected to our SPA application which will start the authentication process.
We use the code flow with PKCE.

We are currently trying to deal with the following problem. To my knowledge, the current system works like this:

- User enters SPA website.
- OIDC component first checks session storage for a valid token
- If the token is valid, the user can continue. If not, the code flow authorization process is started
- If everything is valid and the user is logged on, the silent renew process starts

The problem with this is that when a backoffice user starts a new impersonation process, the old tab might still be open. When you refresh the old tab before the silent renew process has updated the token the page will be reloaded using the old token. However, because the silent renew process starts right after the reload the old token is quickly replaced with the new one (which could be from a completely different user). But, the page was already loaded and is showing data from the old token. This is a big issue for us as we've built a multi-tenant system. 

Is there a way to force the silent renew process to run manually (and wait for it on load)? We want to force the OIDC component to check and possibly get a new token.

Maybe there's something wrong with our setup and that's why this isn't working. If so, please let me know.

I'm hoping someone can help me with this issue.

Thnx!

**Comment by Zetanova** [comment]
It is the same issue as https://github.com/damienbod/angular-auth-oidc-client/issues/763

@florisrobbemont for multi tab logins, u can implement a custom storage and add to the keys  different prefixes to each tab/login

Here is some code blob, that u can use as temple:

**config**
```typescript
export function configureAuth(oidcConfigService: OidcConfigService, userSession: UserSessionService) {
  const config:OpenIdConfiguration = {
    //...
  };
  return async () => {
    await userSession.load();
    return oidcConfigService.withConfig(config);
  };
}

//NgModule.imports
AuthModule.forRoot({
      storage: OidcSessionStorageService
})
```
 
**oidc-session-storage.service.ts**
```typescript
import { Injectable } from '@angular/core';
import { NGXLogger } from 'ngx-logger';
import { AbstractSecurityStorage } from 'angular-auth-oidc-client';
import { UserSessionService } from './user-session.service';

@Injectable()
export class OidcSessionStorageService implements AbstractSecurityStorage {

  constructor(private userSession:UserSessionService, private logger: NGXLogger) { }

  read(key: string) {
    if (!this.userSession.isLoaded) {
      this.logger.warn(`Wanted to read '${key}' but session not loaded`);
      return false;
    }

    const item = this.userSession.get(key);

    if (!item) {
      this.logger.trace(`Wanted to read '${key}' but nothing was found`);
      return null;
    }

    this.logger.trace(`user session storage '${key}' read`);
    return item;
  }
  write(key: string, value: any): boolean {
    if (!this.userSession.isLoaded) {
      this.logger.debug(`Wanted to write '${key}/${value}' but session not loaded`);
      return false;
    }

    const item = this.userSession.get(key);

    if (item == value) {
      this.logger.trace(`user session storage '${key}' unchanged`);
      return true;
    }

    if (!value) {
      this.userSession.delete(key);
      this.logger.trace(`user session storage '${key}' deleted`);
      return true;
    }

    if(key.endsWith('access_token_expires_at')) {
      this.userSession.expiresAt = value;
    } 
    else if(key.endsWith('userData')) {
      this.userSession.sub = value.sub;
      this.userSession.name = value.name;
    }

    this.userSession.set(key, value);
    this.logger.trace(`user session storage '${key}' set`);
    return true;
  }
}
```

**user-session.service.ts**
```typescript
import { Injectable, OnDestroy } from '@angular/core';
import { UserSessionStorageService, UserSessionInfo, UserSessionData } from './user-session-storage.service';
import { BehaviorSubject, ArgumentOutOfRangeError, of, Subscription, Subject } from 'rxjs';
import { map, ignoreElements, switchMap, debounceTime, filter } from 'rxjs/operators';
import { NGXLogger } from 'ngx-logger';

@Injectable({
  providedIn: 'root'
})
export class UserSessionService implements OnDestroy {

  get isLoaded() { return this._info.id >= 0; }

  private _info: UserSessionInfo = {
    id: -1,
    sub: '',
    name: '',
    expiresAt: 0,
    locked: false
  };
  get info() {
    return this._info;
  }
  get name() {
    return this._info.name;
  }
  set name(value: string) {
    this._info.name = value;
  }
  get sub() {
    return this._info.sub;
  }
  set sub(value: string) {
    this._info.sub = value;
  }
  get expiresAt() {
    return this._info.expiresAt;
  }
  set expiresAt(value: number) {
    this._info.expiresAt = value;
  }

  private _data: any = {};

  private _save = new Subject<UserSessionData>();
  private _saveDelayed: Subscription;

  constructor(private storage: UserSessionStorageService, private logger: NGXLogger) {

    //todo fix on session switch
    this._saveDelayed = this._save.pipe(
      debounceTime(100),
      switchMap(n => this.storage.write(n))
    ).subscribe(_ => this.logger.debug("user session saved"));
  }

  ngOnDestroy(): void {
    this._saveDelayed?.unsubscribe();
  }

  async load(id: number = -1) {
    const selection$ = id < 0
      ? this.storage.getSelection().pipe(map(n => n > 0 ? n : 0))
      : of(id);

    let session = await selection$.pipe(
      switchMap(n => this.storage.read(n)),
      map(n => n as UserSessionData)
    ).toPromise();

    if (!session) {
      if (id >= 0) throw new Error(`user session ${id} locked`);

      let max = -1;
      await this.storage.getInfos().forEach(n => {
        if (max < n.id) max = n.id;
      });

      this._info = {
        id: max + 1,
        sub: '',
        name: '',
        expiresAt: 0,
        locked: false
      };
      this._data = {}
      this.logger.debug(`user session ${this._info.id} created`);       
      return;
    }

    this._info = {
      id: session.id,
      sub: session.sub,
      name: session.name,
      expiresAt: session.expiresAt,
      locked: session.nonce != ''
    };

    this._data = session?.data;
    this.logger.debug(`user session ${this._info.id} loaded`);

    await this.storage.select(this._info.id).toPromise();
  }

  save() {
    if (!this.isLoaded)
      throw new Error("user session not loaded");

    const info = this._info;
    const session: UserSessionData = {
      id: info.id,
      sub: info.sub,
      name: info.name,
      expiresAt: info.expiresAt,
      nonce: '',
      data: this._data
    }

    this._save.next(session);
    this.logger.debug(`user session ${this._info.id} save triggered`);
  }

  get(key: string) {
    if (!this.isLoaded)
      throw new Error("user session not loaded");

    return this._data[key];
  }

  set(key: string, value: any) {
    if (!this.isLoaded)
      throw new Error("user session not loaded");

    if (this._data[key] == value)
      return;

    if (value != null) {
      this._data[key] = value;
    } else {
      delete this._data[key];
    }

    this.save();
  }

  delete(key: string) {
    if (!this.isLoaded)
      throw new Error("user session not loaded");

    if (!this._data.hasOwnProperty(key))
      return;

    delete this._data[key];

    this.save();
  }
}


```

**Comment by a-a-k** [comment]
@florisrobbemont hi, could you describe your approach for impersonation implementation?

**Comment by florisrobbemont** [comment]
> @florisrobbemont hi, could you describe your approach for impersonation implementation?

Hey man! Sorry, this is over 2 years ago and we moved on from this project. 
I can remember that we basically triggered the backend to create an authentication ticket for a specific user (while saving the user that started the impersonation process). We used a different cookie-domain for this.
Our backend ran on a seperate domain from the frontend application as well, to make sure two users could be logged on at the same time (one for the backend, and one for the actual frontend application that end-users use)

Hope this helps you further :)
